site/src/routes/pins-add.js

```diff
@@ -3,43 +3,114 @@ import { verifyToken } from '../utils/utils.js'
 import * as PinataPSA from '../pinata-psa.js'
 import { JSONResponse } from '../utils/json-response.js'
 import * as nfts from '../models/nfts.js'
+import * as cluster from '../cluster.js'
 
 /**
  * @param {FetchEvent} event
  * @returns {Promise<JSONResponse>}
  */
-export async function pinsAdd (event) {
+export async function pinsAdd(event) {
   const result = await verifyToken(event)
   if (!result.ok) {
     return HTTPError.respond(result.error)
   }
   const { user, tokenName } = result
   const pinData = await event.request.json()
-  pinData.meta = pinData.meta || {}
-  pinData.meta.sub = user.sub
 
-  const res = await PinataPSA.pinsAdd(pinData)
-  if (!res.ok) {
-    return PinataPSA.parseErrorResponse(res.error)
+  if (typeof pinData.cid !== 'string') {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'invalid CID' } },
+      { status: 400 }
+    )
+  }
+  if (pinData.name && typeof pinData.name !== 'string') {
+    return new JSONResponse(
+      { error: { reason: 'INVALID_PIN_DATA', details: 'invalid name' } },
+      { status: 400 }
+    )
+  }
+
+  const name = pinData.name
+
+  let meta
+  if (pinData.meta) {
+    if (typeof pinData.meta !== 'object' || Array.isArray(pinData.meta)) {
+      return new JSONResponse(
+        { error: { reason: 'INVALID_PIN_DATA', details: 'invalid metadata' } },
+        { status: 400 }
+      )
+    }
+    meta = Object.fromEntries(
+      Object.entries(pinData.meta).filter(([, v]) => typeof v === 'string')
+    )
   }
 
-  const pinStatus = res.value
+  /** @type import('@nftstorage/ipfs-cluster').PinResponse | undefined */
+  let pin
+  /** @type import('../pinata-psa').Status */
+  let status = 'pinning'
+  let size = 0
+  try {
+    pin = await cluster.allocation(pinData.cid)
+  } catch (err) {
+    // if 404 we can continue, we have not pinned this CID
+    if (!err.response || err.response.status !== 404) {
+      throw err
+    }
+  }
 
+  if (pin) {
+    // if we cached the size then use it
+    if (pin.metadata && pin.metadata.size) {
+      size = parseInt(pin.metadata.size)
+    }
+    // we have this pin already allocated in our cluster, get the status...
+    const { peerMap } = await cluster.status(pin.cid)
+    status = cluster.toPSAStatus(Object.values(peerMap))
+    // if this failed to pin, try again
+    if (status === 'failed') {
+      await cluster.recover(pinData.cid)
```

Comment:
Firstly, because we're passing a CID this'll only recover a single pin. Recovery just re-attempts to pin the data if it has an error status. For example it may have tried for 24 hours (or whatever the timeout is) and then given up.

So this is just incase we tried to pin a thing before but it failed and we're attempting to pin the same thing again.